shared/src/main/scala-2.13.0-pre-47ffa9c-SNAPSHOT/scala/util/parsing/input/PagedSeq.scala

```diff
@@ -0,0 +1,274 @@
+   *  a line separator `\n` between any two lines in the sequence.
+   */
+  def fromLines(source: Iterator[String]): PagedSeq[Char] = {
+    var isFirst = true
+    fromStrings(source map { line =>
+      if (isFirst) {
+        isFirst = false
+        line
+      } else "\n"+line
+    })
+  }
+
+  /** Constructs a paged character sequence from a line iterable
+   *  Lines do not contain trailing `\n` characters; The method inserts
+   *  a line separator `\n` between any two lines in the sequence.
+   */
+  def fromLines(source: Iterable[String]): PagedSeq[Char] =
+    fromLines(source.iterator)
+
+  /** Constructs a paged character sequence from an input reader
+   */
+  def fromReader(source: JReader): PagedSeq[Char] =
+    new PagedSeq(source.read(_: Array[Char], _: Int, _: Int))
+
+  /** Constructs a paged character sequence from an input file
+   */
+  def fromFile(source: File): PagedSeq[Char] =
+    fromReader(new FileReader(source))
+
+  /** Constructs a paged character sequence from a file with given name
+   */
+  def fromFile(source: String): PagedSeq[Char] =
+    fromFile(new File(source))
+
+  /** Constructs a paged character sequence from a scala.io.Source value
+   */
+  def fromSource(source: scala.io.Source) =
+    fromLines(source.getLines())
+}
+
+
+import PagedSeq._
+
+/** An implementation of lazily computed sequences, where elements are stored
+ *  in "pages", i.e. arrays of fixed size.
+ *
+ *  A paged sequence is constructed from a function that produces more elements when asked.
+ *  The producer function - `more`, is similar to the read method in java.io.Reader.
+ *  The `more` function takes three parameters: an array of elements, a start index, and an end index.
+ *  It should try to fill the array between start and end indices (excluding end index).
+ *  It returns the number of elements produced, or -1 if end of logical input stream was reached
+ *  before reading any element.
+ *
+ *  @tparam T     the type of the elements contained in this paged sequence, with an `ClassTag` context bound.
+ *
+ *  @author Martin Odersky
+ *  @define Coll `PagedSeq`
+ *  @define coll paged sequence
+ *  @define mayNotTerminateInf
+ *  @define willNotTerminateInf
+ */
+class PagedSeq[T: ClassTag] protected(
+  more: (Array[T], Int, Int) => Int,
+  first1: Page[T],
+  start: Int,
+  end: Int)
+extends scala.collection.AbstractSeq[T]
+   with scala.collection.IndexedSeq[T]
+{
+  def this(more: (Array[T], Int, Int) => Int) = this(more, new Page[T](0), 0, UndeterminedEnd)
+
+  // Members declared in scala.collection.Seq
+  def iterableFactory: collection.SeqFactory[collection.IndexedSeq] = collection.IndexedSeq
+
+  // Members declared in scala.collection.IterableOps
+  protected[this] def fromSpecificIterable(coll: collection.Iterable[T]): collection.IndexedSeq[T] = iterableFactory.from(coll)
+  protected[this] def newSpecificBuilder(): collection.mutable.Builder[T, collection.IndexedSeq[T]] = iterableFactory.newBuilder()
```

Comment:
The only difference with the 2.11 and 2.12 implementation is these 3 lines above that have been added.

Reply:
What do you think about moving those methods to a separate trait in a separate source file? In versions up to 2.12 the source file would be empty. Then we can avoid duplicating the entire file.

Reply:
Done!